@alloy => [Directive] Implement a canonicalResource directive

mzikherman opened the pull request:
This implements `@canonicalResource`, which works like the following:

```
query {
  artist(id: "andy-warholl")  @canonicalResource {
    ... on Artist {
      id 
      alternate_names
    }
  }
}
```

This gives the following in `extensions`:

```
"canonicalResource": {
  "httpStatusCode": 404
}
```

Some quirks I'm discovering so far:
- When the top-level field of a query is canonical, this works. However, if (for some reason), the `alternate_names` field was canonical and threw an error (from the example above), the full path to the error would be: `["artist", "alternate_names"]`. From the visitor pattern, I can get `alternate_names` as the field which has the directive, but I can't seem to build the full path which would be needed to match against the error. I experimented with augmenting the visitor with `SelectionSet` and recursive enter/leaving to try to build up a fuller path.

- This code runs _before_ the error handler/formatter. The latter is what munges errors in order to propagate any HTTP status codes. That needs to be duplicated here (since this code wants to propagate those).

mzikherman commented on package.json:
```diff
@@ -70,7 +70,7 @@
     "graphql-middleware": "1.2.6",
     "graphql-relay": "0.5.4",
-    "graphql-tools": "4.0.4",
+    "graphql-tools": "4.0.5",
```
This upgrade is what allows directives to be properly merged in.

mzikherman commented on _schema.graphql:
```diff
@@ -1,3 +1,5 @@
+directive @canonicalResource on FIELD
```
🎂

alloy replied:
✨

mzikherman replied:
We should add this in v2 as well.

damassi commented on src/index.ts:
```diff
@@ -5,6 +5,7 @@ import { applyMiddleware as applyGraphQLMiddleware } from "graphql-middleware"
 import bodyParser from "body-parser"
 import config from "./config"
 import cors from "cors"
+import { get } from "lodash"
```
@mzikherman - this is now safe to add as it's been moved to stage 3: https://github.com/tc39/proposal-optional-chaining

damassi replied:
(Also this https://github.com/tc39/proposal-nullish-coalescing, which is a slightly different take)

damassi replied:
Ah, actually -- the TypeScript compiler isn't yet there.

alloy commented on src/lib/loaders/getCanonicalResourceDirectiveField.ts:
```diff
@@ -2,31 +2,23 @@ import { FieldNode, visit, BREAK } from "graphql"
+      },
+      leave() {
+        path.pop()
+      },
```
Keep a stack…

mzikherman replied:
Sweet, yea this makes perfect sense.

alloy replied:
Btw, I actually expect the error path to use field aliases when those exist, whereas here I was only using the real field name, so that needs fixing too.

mzikherman replied:
Yea, was wondering about stitching. Errors occurring using `ecommerceOrder` have that as the path (and `alias: undefined`), but the error reporting path uses `order`.

alloy replied:
(Will respond to the stitched errors issue outside of this conversation.)

alloy commented on src/lib/loaders/getCanonicalResourceDirectiveField.ts:
```diff
@@ -2,31 +2,23 @@ import { FieldNode, visit, BREAK } from "graphql"
       if (node.name.value === "canonicalResource") {
-        const containingField = ancestors.slice(-1)[0] as FieldNode
-        field = containingField.name.value
         return BREAK
```
…then return the current state of the stack when we hit a directive.

alloy commented on src/lib/loaders/getCanonicalResourceDirectiveField.ts:
```diff
@@ -2,31 +2,23 @@ import { FieldNode, visit, BREAK } from "graphql"
   })
 
-  return field
+  return path
```
One thing left to do is to either return an empty array or `null`.

mzikherman replied:
ya, i would return `null` in the event the directive isnt used.

alloy commented on src/schema/v1/artist/index.ts:
```diff
@@ -93,7 +93,12 @@ export const ArtistType = new GraphQLObjectType<any, ResolverContext>({
+        resolve: () => {
+          throw new Error("oh noes")
+        },
+      },
```
Sigh.

alloy commented on src/index.ts:
```diff
@@ -212,16 +212,13 @@ function startApp(appSchema, path: string) {
             )
 
-            if (result.errors && result.errors.length) {
+            if (canonicalFieldPath && result.errors && result.errors.length) {
```
This check actually currently makes no sense, as `getCanonicalResourceDirectiveForField` will always return an array.

mzikherman replied:
Cool, so with the path matching, that leaves cleaning up the error munging/introspection. This was a fast take, I'll take another pass and consolidate with the error formatter.

alloy commented on src/lib/loaders/getCanonicalResourceDirectiveField.ts:
```diff
@@ -0,0 +1,24 @@
+import { FieldNode, visit, BREAK } from "graphql"
+
+export const getCanonicalResourceDirectiveForField = (
+  fieldNode: FieldNode
```
We’re not actually passing in a field node, are we?

mzikherman replied:
Ah yea, it's a `DocumentNode`.

mzikherman commented:
Updated with some refactoring and consolidation of error munging between the error and extensions middleware. Also added a validation that only one `@canonicalResource` directive appears.

New quirk:
  - an error during stitching, like:
```
commerceOrder(id:"123123") @canonicalResource {
    mode
}
```

winds up having an error path of `order`, and not `commerceOrder`.

mzikherman commented on src/extensions/canonicalResourceDirectiveExtension.ts:
```diff
@@ -0,0 +1,28 @@
+            canonicalExtensions["canonicalResource"] = {
+              httpStatusCode,
+            }
+            return
```
Was trying to break early, but I'm realizing you can't from `forEach`! Will rewrite.

alloy replied:
Ahhh, yeah I should have realized when I wrote my above comment. Use something like `Artay.prototype.some()` perhaps?

mzikherman commented on src/lib/getCanonicalResourceDirectiveField.ts:
```diff
@@ -0,0 +1,26 @@
+  visit(documentNode, {
+    Field: {
+      enter(node) {
+        const name = get(node, "alias.value") || node.name.value
```
over `node.alias.value`? No reason! Just wanted to be extra safe.

mzikherman replied:
Oh, `alias` could be `undefined`, based on the typings for it.

alloy replied:
Right, but with `get` you loose type checking and is less performant. My opinion is that it should only ever be used for inputs you can’t trust, whereas this is an API that isn’t suddenly changing at runtime.

A version without `get` could be even more concise:

```ts
const name = (node.alias || node.name).value
```

mzikherman replied:
Ah, I like this.

alloy commented:
So yeah, those stitched errors and their paths. That’s technically a bug in the schema stitching code, in my opinion, but unsure immediately how to address. Perhaps first check if an issue exists for this over in the graphql-tools repo?

alloy commented on src/extensions/canonicalResourceDirectiveExtension.ts:
```diff
@@ -6,21 +6,20 @@ export const canonicalResourceDirectiveExtension = (documentAST, result) => {
-        const errors = flattenErrors(e)
+    const errors = result.errors
+      .filter(e => isEqual(e.path, canonicalFieldPath))
+      .pop()
```
Use `Array.prototype.find()` here, which, like `some`, will stop as soon as a match is found.

mzikherman commented:
In terms of getting this merged (and iterated on!), I think getting the current code review comments addressed and adding some specs might do it?

The stitching issue remains, but I added a patch that adds the directive to V2 as well.

alloy commented:
Agreed! Exciting times to be alive 👨‍🎤🤘

Perhaps the last thing to consider is a different name for the directive? I liked your ‘field’ over ‘resource’ suggestion. What about ‘principal’ over ‘canonical’?

mzikherman commented:
Updated with specs, this should be mergeable.

Kept the naming as `canonicalResource`, I think the debate was between that and `principalField`.

I'm not very strongly opinionated on that to be honest. I don't mind keeping it as-is.

alloy commented on src/schema/v1/__tests__/canonicalResourceDirective.test.ts:
```diff
@@ -0,0 +1,33 @@
+      queryToAst(query),
+      result
+    )
+    expect(extensions).toEqual({ canonicalResource: { httpStatusCode: 404 } })
```
Perhaps this test would be slightly better if you would also raise an error from another field but with a different status code.

alloy commented on _schema.graphql:
```diff
@@ -1,3 +1,5 @@
+directive @canonicalResource on FIELD
```
Can we have documentation for directives?

alloy replied:
@mzikherman bump

mzikherman replied:
double-checking...

mzikherman commented:
Alright, renamed to `principalField`. Will #mergeongreen

alloy commented:
![](https://www.flavourmag.co.uk/files/2017/08/epic-dinner-party.gif)

kierangillen commented on src/index.ts:
```diff
@@ -193,24 +195,34 @@ function startApp(appSchema, path: string) {
+              ? fetchLoggerRequestDone(requestID)
+              : {}
+
+            return { ...principalFieldExtensions, requestLoggerExtensions }
```
Pairing with @ds300 on the graphql issue we're seeing now in Emission -- Looks like this needs to be `return { ...principalFieldExtensions, ...requestLoggerExtensions }` as `res.json.extensions` is missing the `requests` object.
